File: skills/playwright/src/commands/content.rs
use crate::browser::BrowserSession;

pub async fn content(
    session: &mut BrowserSession,
    format: &str,
) -> anyhow::Result<serde_json::Value> {
    let content = if format == "html" {
        session.page.content().await?
    } else {
        session
            .page
            .evaluate("document.body.innerText")
            .await?
            .into_value::<String>()
            .unwrap_or_default()
    };

    Ok(serde_json::json!({
        "content": content,
        "url": session.url().await,
    }))
}

pub async fn text(
    session: &mut BrowserSession,
    selector: &str,
) -> anyhow::Result<serde_json::Value> {
    let element = session
        .page
        .find_element(selector)
        .await
        .map_err(|e| anyhow::anyhow!("Element not found: {e}"))?;

    let text = element
        .inner_text()
        .await?
        .unwrap_or_default();

    Ok(serde_json::json!({
        "text": text.trim(),
        "url": session.url().await,
    }))
}

pub async fn snapshot(
    session: &mut BrowserSession,
) -> anyhow::Result<serde_json::Value> {
    // Use CDP Accessibility.getFullAXTree for accessibility snapshot
    let snapshot = session
        .page
        .evaluate(
            r#"
            (function() {
                function getAccessibilityTree(node, depth) {
                    if (depth > 10) return '';
                    let result = '';
                    const indent = '  '.repeat(depth);
                    const role = node.getAttribute && node.getAttribute('role') || node.nodeName.toLowerCase();
                    const text = node.textContent ? node.textContent.trim().substring(0, 100) : '';
                    const ariaLabel = node.getAttribute && node.getAttribute('aria-label') || '';

                    if (role !== '#text' && role !== 'script' && role !== 'style') {
                        let line = `${indent}${role}`;
                        if (ariaLabel) line += ` "${ariaLabel}"`;
                        else if (text && !node.children.length) line += ` "${text}"`;
                        result += line + '\n';
                    }

                    for (const child of (node.children || [])) {
                        result += getAccessibilityTree(child, depth + 1);
                    }
                    return result;
                }
                return getAccessibilityTree(document.body, 0);
            })()
            "#,
        )
        .await?
        .into_value::<String>()
        .unwrap_or_default();

    Ok(serde_json::json!({
        "snapshot": snapshot,
        "url": session.url().await,
    }))
}

File: skills/playwright/src/commands/content.ts
import { getBrowser } from "../browser.js";

export interface ContentResult {
	success: boolean;
	content: string;
	url: string;
	timestamp: string;
	error?: string;
}

export interface TextResult {
	success: boolean;
	text: string;
	url: string;
	timestamp: string;
	error?: string;
}

export interface SnapshotResult {
	success: boolean;
	snapshot: string;
	url: string;
	timestamp: string;
	error?: string;
}

export async function content(
	format: "text" | "html" = "text",
): Promise<ContentResult> {
	try {
		const { page } = await getBrowser();

		let contentStr: string;
		if (format === "html") {
			contentStr = await page.content();
		} else {
			contentStr = await page.evaluate(() => document.body.innerText);
		}

		return {
			success: true,
			content: contentStr,
			url: page.url(),
			timestamp: new Date().toISOString(),
		};
	} catch (err) {
		const error = err instanceof Error ? err.message : String(err);
		return {
			success: false,
			content: "",
			url: "",
			timestamp: new Date().toISOString(),
			error,
		};
	}
}

export async function text(selector: string): Promise<TextResult> {
	try {
		const { page } = await getBrowser();
		const element = await page.locator(selector).first();
		const textContent = (await element.textContent()) || "";

		return {
			success: true,
			text: textContent.trim(),
			url: page.url(),
			timestamp: new Date().toISOString(),
		};
	} catch (err) {
		const error = err instanceof Error ? err.message : String(err);
		return {
			success: false,
			text: "",
			url: "",
			timestamp: new Date().toISOString(),
			error,
		};
	}
}

export async function snapshot(): Promise<SnapshotResult> {
	try {
		const { page } = await getBrowser();
		// Use ariaSnapshot for accessibility tree (Playwright 1.49+)
		const tree = await page.locator("body").ariaSnapshot();

		return {
			success: true,
			snapshot: tree,
			url: page.url(),
			timestamp: new Date().toISOString(),
		};
	} catch (err) {
		const error = err instanceof Error ? err.message : String(err);
		return {
			success: false,
			snapshot: "",
			url: "",
			timestamp: new Date().toISOString(),
			error,
		};
	}
}
